### PR TITLE
Add docker support

### PR DESCRIPTION
**Pull request by ezmi234**
Added Docker support. The RISC-V container image was built with the latest updated dependencies to ensure a smoother setup and execution.

Note:
Please review the create_predictor.py file. Within the Docker environment, execution proceeds without BTBEntries since the TournamentBP class does not support it as a parameter. To preserve lab configuration compatibility, this behavior is managed using an environment variable, so no existing setup should be affected.
It was test only on macos with apple silicon.

### CODE REVIEW

File: setup_default [hunk 1]
```diff
@@ -5,8 +5,13 @@ export WORK_DIR=$(pwd)
 ###############################################
 #########   CROSS COMPILER RISCV  #############
 ###############################################
-export CC="/opt/riscv-2023.10.18/bin/riscv64-unknown-elf-gcc"
-export CC_INSTALLATION_PATH="/opt/riscv-2023.10.18/"
+if [[ -n "$IS_DOCKER" ]]; then
```
Comment (franout) [inline setup_default:8]: I would rather use a different setup_default such as ```setup_default.docker```

File: gem5/create_predictor.py [hunk 2]
```diff
@@ -29,18 +30,19 @@ def create_LocalBP():
...
 def create_TournamentBP():
-    # TOURNAMENT BP it inherits BranchPredictor 
+    # TOURNAMENT BP it inherits BranchPredictor
     pred = TournamentBP()
     pred.localPredictorSize = 32
     pred.localHistoryTableSize = 256
     pred.globalPredictorSize = 64
     pred.choicePredictorSize = 64
-    pred.BTBEntries = 256
+    if os.getenv('IS_DOCKER') != 'true':
```
Comment (franout) [inline gem5/create_predictor.py:44]: i do not understand why docker is affecting the gem5 predictors.
This is something that should not happen

File: gem5/create_predictor.py
Comment (franout) [inline gem5/create_predictor.py:1]: @ezmi234  please adjust the spaces compared to the main. It is bad to have diffs based on spaces

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      # - /tmp/.X11-unix:/tmp/.X11-unix  # Share the X11 socket
+      - ./results:/opt/konata/results    # Mount results directory
+    image: ezmi2345/konata:latest
+
+  riscv:
+    build:
+      context: .
+      dockerfile: Dockerfile.riscv
+    environment:
+      - IS_DOCKER=true
```
Comment (franout) [inline docker-compose.yml:18]: see setup default commments

File: README.md [hunk 4]
```diff
@@ -15,9 +15,11 @@ This README provides an environment for simulating a program on a parametrizable
     - [Installing Konata, the Pipeline Visualizer](#installing-konata-the-pipeline-visualizer)
   - [HOWTO - Simulate a Program](#howto---simulate-a-program)
   - [HOWTO - Visualize the Pipeline with Konata](#howto---visualize-the-pipeline-with-konata)
+  - [Installation with Docker](#installation-with-docker)
```
Comment (franout) [inline README.md:18]: it is probably more logical to add this in the installation section

File: Dockerfile.riscv [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    rm -rf /var/lib/apt/lists/*
+
+# Clone the RISC-V GNU Toolchain
+RUN git clone --recursive https://github.com/riscv-collab/riscv-gnu-toolchain /opt/riscv-gnu-toolchain
+
+# Clone gem5
+RUN git clone --recursive https://github.com/gem5/gem5 /opt/gem5
+
+# Build the 64-bit RV64GC toolchain with glibc
+RUN cd /opt/riscv-gnu-toolchain && ./configure --prefix=/opt/riscv
```
Comment (franout) [inline Dockerfile.riscv:53]: the toolchain must be compiled with --enable-multilib. Please see #4

File: Dockerfile.riscv [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    cmake \
+    libglib2.0-dev \
+    libslirp-dev && \
+    rm -rf /var/lib/apt/lists/*
+
+# Clone the RISC-V GNU Toolchain
+RUN git clone --recursive https://github.com/riscv-collab/riscv-gnu-toolchain /opt/riscv-gnu-toolchain
+
+# Clone gem5
+RUN git clone --recursive https://github.com/gem5/gem5 /opt/gem5
```
Comment (franout) [inline Dockerfile.riscv:50]: be careful. The simulation environment works for the gem5 v22.1.0.0. You are missing the checkout command

File: Dockerfile.konata
Comment (franout) [inline Dockerfile.konata:1]: why do you need a dockerfile for Konata?
as far i remember, ready to ship executables for konata are already available.

File: DockerDocs.md
Comment (franout) [inline DockerDocs.md:1]: duplicate in README.md?

File: README.md [hunk 7]
```diff
@@ -134,10 +136,46 @@ To visualize the pipeline, follow these steps:
...
+Konata is available in the `konata` container and can be run independently from the gem5 simulation.
+
+### Running the RISC-V Cross-Compiler
+To run a program using the cross-compiler and gem5, execute:
+```bash
+docker exec -it ase_riscv_gem5_sim-riscv-1 bash -c "./simulate.sh -i ./programs/fibonacci/ -nogui"
+```
+
+### Using Konata
+Konata is used to open and visualize `.out` files generated in the `results` folder. Before starting the container:
```
Comment (franout) [inline README.md:164]: i suggest you to explain better this part.

File: README.md [hunk 8]
```diff
@@ -134,10 +136,46 @@ To visualize the pipeline, follow these steps:
...
+```bash
+docker compose up -d
+```
+
+### Usage
+The setup includes the gem5 simulator and a RISC-V cross-compiler within the `riscv` container. You can attach to this container and run simulations as needed. Simulation results are saved in the `results` directory, and you can view the generated `.out` files using the Konata visualization tool.
+
+Konata is available in the `konata` container and can be run independently from the gem5 simulation.
+
+### Running the RISC-V Cross-Compiler
```
Comment (franout) [inline README.md:157]: this is incorrect!
the cross compiler is a part of the flow! please read carefully the README.md and the slides.

File: README.md [hunk 9]
```diff
@@ -134,10 +136,46 @@ To visualize the pipeline, follow these steps:
...
+### Starting the Environment
+To start the Docker environment, use:
+```bash
+docker compose up -d
+```
+
+### Usage
+The setup includes the gem5 simulator and a RISC-V cross-compiler within the `riscv` container. You can attach to this container and run simulations as needed. Simulation results are saved in the `results` directory, and you can view the generated `.out` files using the Konata visualization tool.
+
+Konata is available in the `konata` container and can be run independently from the gem5 simulation.
```
Comment (franout) [inline README.md:155]: i am not getting the point. Why do you want to use a container for konata?

**Review by franout — CHANGES_REQUESTED** [review]
@ezmi234  thank you for the contribution.
few comments:
- [ ]  please understand the **simulation** flow.
- [ ]  konata is already available in precompiled binaries
- [ ]  i would add an image for the granflow of the container and how things work, somebody could get lost.
- [ ]  i see you used ubuntu22.04, does it work for windows? I think linux-based OSes and mac are fine, the big issue is for windows users.

after you fix those, we can see if it works and merge

**Comment by franout** [comment]
@ezmi234  do we need to considered the merge as discontinued?

**Comment by ezmi234** [comment]
No, ci sto ancora lavorando(oggi ho finito le build delle immagini), domani le faccio vedere in laboratorio. Dovrei aver risolto tutto il difficile

**Pull request closed without merging**